starlite/contrib/sqlalchemy/repository.py

```diff
@@ -0,0 +1,441 @@
+            return instance
+
+    async def get_one_or_none(self, **kwargs: Any) -> ModelT | None:
+        """Get instance identified by `item_id`.
+
+        Args:
+            item_id: Identifier of the instance to be retrieved.
+
+        Returns:
+            The retrieved instance.
+
+        Raises:
+            RepositoryNotFoundException: If no instance found identified by `item_id`.
+        """
+        with wrap_sqlalchemy_exception():
+            select = self._filter_select_by_kwargs(select=self.select, **kwargs)
+            instance = (await self._execute(select)).scalar_one_or_none()
+            instance = self.check_not_found(instance)
+            self.session.expunge(instance)
+            return instance
+
+    async def count(self, *filters: FilterTypes, **kwargs: Any) -> int:
+        """Get the count of records returned by a query.
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            Count of records returned by query, ignoring pagination.
+        """
+        select = self.select.with_only_columns(
+            sql_func.count(  # pylint: disable=not-callable
+                self.model_type.id,  # type:ignore[attr-defined]
+            ),
+            maintain_column_froms=True,
+        ).order_by(None)
+        select = self._apply_filters(*filters, apply_pagination=False, select=select)
+        select = self._filter_select_by_kwargs(select, **kwargs)
+        results = await self._execute(select)
+        return results.scalar_one()  # type: ignore
+
+    async def list(self, *filters: FilterTypes, **kwargs: Any) -> abc.Sequence[ModelT]:
+        """Get a list of instances, optionally filtered.
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            The list of instances, after filtering applied.
+        """
+        select = self._apply_filters(*filters, select=self.select)
+        select = self._filter_select_by_kwargs(select, **kwargs)
+
+        with wrap_sqlalchemy_exception():
+            result = await self._execute(select)
+            instances = list(result.scalars())
+            for instance in instances:
+                self.session.expunge(instance)
+            return instances
+
+    async def list_and_count(
+        self,
+        *filters: FilterTypes,
+        **kwargs: Any,
+    ) -> tuple[abc.Sequence[ModelT], int]:
+        """List records with total count.
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            Count of records returned by query, ignoring pagination.
+        """
+        select = self.select.add_columns(
+            over(
+                sql_func.count(  # pylint: disable=not-callable
```

Comment:
is this type of thing well supported across different backends?

Reply:
Supported by SQLite, MySQL/Maria, Oracle, MS and Postgres. Other I don't know, but that's good enough I think.

Reply:
Yeah, analytic window functions are standard now for most engines.  This is actually the reason why I left it like this instead of breaking it into 2 queries.  However, if we run into issues, we can add the fallback to use 2 queries for this.